Kernel/Process.cpp

```diff
@@ -760,21 +760,44 @@ pid_t Process::sys$fork(RegisterState& regs)
     return child->pid();
 }
 
+void Process::kill_threads_except_self()
+{
+    InterruptDisabler disabler;
+
+    if (m_thread_count == 1)
+        return;
+
+    for_each_thread([&](Thread& thread) {
+        if (&thread == Thread::current
+                || thread.state() == Thread::State::Dead
+                || thread.state() == Thread::State::Dying)
+            return IterationDecision::Continue;
+
+        // At this point, we have no joiner anymore
+        thread.m_joiner = nullptr;
+        thread.set_should_die();
+
+        if (thread.state() != Thread::State::Dead)
+            thread.set_state(Thread::State::Dying);
+
+        return IterationDecision::Continue;
+    });
+
+    big_lock().clear_waiters();
+}
+
 int Process::do_exec(NonnullRefPtr<FileDescription> main_program_description, Vector<String> arguments, Vector<String> environment, RefPtr<FileDescription> interpreter_description)
 {
     ASSERT(is_ring3());
+
     auto path = main_program_description->absolute_path();
     dbg() << "do_exec(" << path << ")";
-    // FIXME(Thread): Kill any threads the moment we commit to the exec().
-    if (thread_count() != 1) {
-        dbgprintf("Gonna die because I have many threads! These are the threads:\n");
-        for_each_thread([](Thread& thread) {
-            dbgprintf("Thread{%p}: TID=%d, PID=%d\n", &thread, thread.tid(), thread.pid());
-            return IterationDecision::Continue;
-        });
-        ASSERT(thread_count() == 1);
-        ASSERT_NOT_REACHED();
-    }
+
+    kill_threads_except_self();
```

Comment:
The FIXME that you are removing says:
>Kill any threads the moment we commit to the exec().

At this point, we have not yet committed to the exec! It can still fail and return with an error.
In those cases, we should not kill any threads.

There's a comment further down where we actually commit:
```
        // NOTE: At this point, we've committed to the new executable.
```

Reply:
Did it there because I used to have a crash, but I commented at the end of the PR more details.

We can be interrupted (and it happens every time with >100 threads) in `do_exec` after the `move(m_regions)` part, and any other thread running from that point will crash if a syscall happens. Fixed it another way (see the new push).